# Benford  for Python -  IP addresses  UNDER DEVELOPMENT

### Current version: 0.1.0.3

# IP Test Case

### Benford for python is a Package in PyPi (as of December 2017). I discovered this package and on Milcent's github page https://github.com/milcent/benford_py and wanted to run a test on some IP addresses.


### I chose to use census data as it is known to be a Benford compliant dataset.  Briefly, Benford's Law deals with the  the frequency distribution of leading digits in many real-life sets of numerical data. The law states that in many naturally occurring collections of numbers, the leading significant digits are not evenly distributed.

### For example, in sets that obey the law, the number 1 appears as the most significant digit about 30% of the time, while 9 appears as the most significant digit less than 5% of the time. By contrast, if the digits were distributed uniformly, they would each occur about 11.1% of the time (Wikipedia).    

### Benford's law is often used in forensic accounting to detect fraud and other anomalies, see (Mark Nigrini and Joseph T. Wells 'Benford's Law: Applications for Forensic Accounting, Auditing and  Fraud Detection)

In [20]:
%matplotlib inline

import numpy as np
import pandas as pd
import pandas_datareader.data as web # Not a dependency, but we'll need it now.

In [21]:
import benford as bf
import os

### Load the Data

In [22]:
good_ips = pd.read_csv('/Users/donajstewart/good_ips_index.csv',encoding='latin-1', index_col="Record")

In [23]:
good_ips.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9867 entries, 1 to 9867
Data columns (total 2 columns):
Server Address    9867 non-null object
first             9867 non-null int64
dtypes: int64(1), object(1)
memory usage: 231.3+ KB


#### The dataframe contains and index column (Record), Server Address and a column with the first digit.  

In [24]:
good_ips.head()

Server Address  first
Record                       
1       34.239.253.124      3
2        151.101.64.70      1
3        151.101.64.70      1
4       173.194.78.189      1
5        172.217.6.142      1

#### There are  9,867 addresses

### First Digits Test

### Let us see if the IP data conforms to Benford's Law 

## Please see error below, I've tried to pass an index parameter in the bf.first_digits function but it does not accept it.

In [25]:
good_ips=bf.first_digits(safe_ip.first, digs=1, confidence=99) 



ValueError: If using all scalar values, you must pass an index